### DISCUSSION TITLE
Pre-PR Checklist and General Workflow

### DISCUSSION THREAD

**Discussion by jhoshiko** [NREL/mappymatch#21] [Ideas]
This pre-PR checklist and workflow is not set in stone. It is meant to provide general guidelines for contributing to the project. We are also open to any feedback and can alter the workflow/checklist to include improvements.

- Select an issue that interests you. We can optionally have a technical discussion or brainstorm ideas for resolving the issue. If you would like to pair up with someone else, feel free to do so! 

- Development work: Create a branch for your issue. When coding your contribution, keep the following in mind:

    1. Functions/Methods/Classes should include docstrings. We are planning to have our docstrings be compatible with [Spinx](https://www.sphinx-doc.org/en/master/), so the docstring format may change. An example of our current docstring structure is attached below. Function/method docstrings should include a function/method description, parameters, and return values. Class docstrings should include a class description, public methods, and instance variables
    2. We use type hints for variables, functions, and return values. Type hints should pass [Mypy](https://github.com/python/mypy) 
        tests. If you need extra information about type hints, feel free to ask one of us or reference the [Mypy cheat sheet](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)
    3. If applicable, your contribution should have unit tests and/or functional tests
    4. Your code has been formatted with the [Black formatter](https://black.readthedocs.io/en/stable/)
    
- We expect that your code will be in scope of the issue. If you think of some additional improvements, submit another issue and 
   create a new branch for it

- Once your code has been submitted, it will be reviewed by two of us on multiple platforms. We are hoping to improve the review process with an automated  CI/CD pipeline which will become part of the review structure once a pipeline is implemented.

**Comment by Rowlando13** [comment]
Docstring example bellow(google flavor): 
```
"""
Make dictionary with desired names as values and codes as keys. 

This row is for text that runs across multiple lines.  

Args:
        df : Dataframe in.
        cols_column : Column names for column that will be processed. 
        desired_names : Series of desired names with type row 
        check_keys_unique : If True checks to make sure keys are 
            unique so don't get overwritten.  

Returns:
        A concise description. 
"""
```
Typing in docs is necessary since sphinx will using the typing hints. 
[Full explanation from Sphinx napoleon extensions](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

**Comment by Rowlando13** [comment]
Based on input from the maintainers and consensus, we are using restructured text with Sphinx. The plan is to set up to autodeploy to readthedocs.
